lib/Dialect/TensorExt/IR/TensorExtAttributes.td

```diff
@@ -0,0 +1,67 @@
+#ifndef LIB_DIALECT_TENSOREXT_IR_TENSOREXTATTRIBUTES_TD_
+#define LIB_DIALECT_TENSOREXT_IR_TENSOREXTATTRIBUTES_TD_
+
+include "lib/Dialect/TensorExt/IR/TensorExtDialect.td"
+
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/DialectBase.td"
+include "mlir/IR/TensorEncoding.td"
+
+class TensorExt_Attr<string name, string attrMnemonic, list<Trait> traits = []>
+    : AttrDef<TensorExt_Dialect, name, traits> {
+  let mnemonic = attrMnemonic;
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def TensorExt_Attr : TensorExt_Attr<"TensorExt", "simd_encoding",
+      [DeclareAttrInterfaceMethods<VerifiableTensorEncoding>]> {
+  let summary = "An attribute describing the SIMD encoding of a tensor.";
+  let description = [{
+    This attribute is used as the encoding attribute on a tensor. It describes
+    the transformations that were applied to an input tensor to pack it into the
+    given tensor.
+
+    The `in` attribute describes the shape of the original tensor.
+    The following transformations are applied to the input tensor.
+
+      1. Padding is applied first. The `padding` attribute is an array with the
+         same size as the input tensor shape. Padding is applied at the end of
+         the array using the `padding_value` attribute (default zero). The
+         result after zero padding should be a power of two.
+
+      2. The padded result is replicated or split to fill the output tensor
+         shape.
+
+    For example,
+
+    ```
+    #encoding = #tensor_ext.simd_encoding<
+      in = [7],
+      padding = [1],
+      padding_value = 0,
+      out = [16],
+    >
+    ```
+
+    may be used on a tensor type like
+
+    ```
+    tensor<1x16xi32, encoding=#encoding>
+    ```
+
+    If the original tensor had values `[1, 2, 3, 4, 5, 6, 7]` then a tensor with
+    this attribute contains the data `[1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6,
+    7, 0]`.
+  }];
+
+  let parameters = (ins
+    ArrayRefParameter<"int64_t">:$in,
+    ArrayRefParameter<"int64_t">:$padding,
+    ArrayRefParameter<"int64_t">:$out,
+    DefaultValuedParameter<"int64_t", "0">:$padding_value
+  );
+
+  let assemblyFormat = "`<` `in` `=` `[` $in `]` `,` `padding` `=` `[` $padding `]` `,` (`padding_value` `=` $padding_value^ `,`)? `out` `=` `[` $out `]` `>`";
```

Comment:
There should be a "struct(params)" type of format here, I think it should be equivalent to this. I believe we use that for lwe parameters and such

Reply:
So struct (params) doesn't work for some reason with arrayref<int>. i'll ask in the MLIR discord.

Reply:
Try DenseI64ArrayAttr maybe? https://github.com/llvm/llvm-project/blob/7476c20c481cbccbdb89139fb94620e083015932/mlir/include/mlir/Dialect/Tosa/IR/TosaOpBase.td#L128

Reply:
That might work if it uses llvm::arrayref instead of smallector - by default the ArrayRefParameter is a SmallVector whose default parser expects no delimeter on the lists (so the list is like `in = 1, 2, 3`), which makes the comma-separated keyword for struct parsing very confused with the keywords or the end of the list. I'll try tomorrow AM!

Reply:
yeah all the ML ops have this similar notation, and I don't think they use custom parsers, so we should try to match what they do if possible.